### ISSUE TITLE
Unexpected CPU usage from af_is_real

### ISSUE THREAD

**Issue by ebarjou** [arrayfire/arrayfire#2905]
Hello, not sure if this is a bug or a misunderstanting from my part :
Using unified backend 3.7.1 with CUDA, the thread running this function is always at 100% CPU usage and the execution time is far above what I expected compared to other calculations I do in the same program :

```
af::array getPM(int index, int layer, bool interp = false) {
	float zCoord = _volMin.y() + layer*_vx;
	af::array layerCoords = af::join(1, 
		_X,
		af::tile(af::array(1, 1, &zCoord), _pw*_pw),
		_Z,
		_W
	);
	
	af::array p = af::matmul(layerCoords, _fm.slice(index)); // 4xN mult by 4x4 matrix
	af::eval(p);
	af::array U = (p.col(0)*_pw)/(2.0f*p.col(3))+_pw/2.0f;
	af::array V = (p.col(1)*_ph)/(2.0f*p.col(3))+_ph/2.0f;
	
	U = af::round(U).as(af::dtype::s32);
	V = af::round(V).as(af::dtype::s32);
	af::array R = !(U<0.0f||U>=float(_pw)||V<0.0f||V>=float(_ph));
	af::array B = af::range(af::dim4(_pw*_pw+1), 0, af::dtype::s32);
	af::array A = U*_ph+V;
	af::replace(A, R, 0.0f);
	af::array W = _layerCoordsW;
	af::replace(W, R, 0.0f);
	
	af::array pm = af::sparse(_pw*_pw, _ph*_pw, W, B, A, af::storage::AF_STORAGE_CSR);
	af::eval(pm);
	return pm;
}
```
All array used are f32.

Using a profiler, "af_is_real" represent consistently more than 99% of the time spent in this function :

```
af_is_real		208.48s	99.87%
__security_check_cookie	0.07s	0.03%
af::array::~array	0.07s	0.03%
af::tile		0.04s	0.02%
ILT+1410(__security_check_cookie)	0.02s	0.01%
af::matmul		0.02s	0.01%
af::array::array<float>	0.02s	0.01%
af::array::slice	0.02s	0.01%
af::operator+		0.02s	0.01%
```
And af_is_real spent most of its time in af::constant<short> and calling itself (?)

```
af_is_real	387.42s	63.20%	af
af::constant<short>	224.67s	36.65%
malloc_base	0.09s	0.01%
af_abs	0.06s	0.01%
malloc	0.06s	0.01%
af_neq	0.06s	0.01%
```

I can't figure out what cause this. Is it that the evals make the thread wait under one of those function, or is there another issue ?

**Comment by 9prady9** [comment]
`isReal` is internally used by sort and unary operations(which is only `af::round` in the code shared), it shouldn't occupy too much of CPU usage especially since it doesn't require any computation on GPU/CPU to tell if `af::array` is real or complex valued.

I can't say for certain if it is eval or something else. But why do you have manual eval calls ? Are you facing any issues  otherwise ? It is often best to leave jit eval to automatic heuristics unless your use case is showing performance degradataion.

Also, what was the tool you used to get these usage statistics ? It would also help if you can reproduce the problem in standalone program so that we can use that code to reproduce and debug any issues on our end more easily.

**Comment by ebarjou** [comment]
I added the eval after seeing that my bottleneck was the CPU core that run this function to see if waiting for the GPU reduced the CPU usage.
I tried without the evals, with and without a sync before the return, the results are exactly the same.

The profiler I used was Very Sleepy CS, but trying again with SuperLuminal allowed to more precisely spot the issue.
Most of the CPU time was spent on the join line, and more precisely from the array<float> inside. Changing this to af::constant effecively fixed the issue.

Not sure if this is still an issue, It look like af::array is called for each element of the tile (I'm not very familiar with profiler tools, but I doubt it's called only once considering the time spent on it !)

**Comment by 9prady9** [comment]
I am not sure where exactly the problem is but profilers usually run the code multiple times to gather the statistics.

The tools you mentioned suggests you are using Windows. I haven't used either of those so far. Did you try the profiler that comes with Visual Studio ? If you can provide a standalone program(code) that I can run, I can try to profile it and see the results on my end.

**Comment by ebarjou** [comment]
I don't have VS installed, but I tried to make a program that replicate the issue.

I execute the following function 100 times, taking care of doing the warmup in a different function : 
```
void computation() {
    af::array A = af::join(1,
        B, C, D, af::tile(af::array(1, &a), dim)
    );
    af::array R = af::matmul(A, E);
}
```
Profiling show that af::array<float> take 96% of the time spent in this function (With AF_SYNCHRONOUS_CALLS = 1 to make sure the matmul is not executed after the function return)

I join a zip with this test program :
[src.zip](https://github.com/arrayfire/arrayfire/files/4681890/src.zip)

**Comment by 9prady9** [comment]
![Issue2905](https://user-images.githubusercontent.com/3270458/82901855-2c869500-9f7c-11ea-8f86-3493ab4e16e6.png)

In my debugging run with CPU profiling enabled, the program hardly reaches 50% of CPU utilization for consistent period of time.

**Comment by ebarjou** [comment]
I installed Visual Studio and tested with its profiler, and it look like the CPU usage reported by VS is for all CPU core, so I'd guess you have a 4 core CPU, with one core at 100%, resulting in ~25% of CPU usage.

It report me around 10% of CPU usage, but only one out of my 6/12 cores is used when running the exemple :
![image](https://user-images.githubusercontent.com/17290883/82909807-0f48cb00-9f6a-11ea-81ff-3c05563b2add.png)

**Comment by 9prady9** [comment]
Can you please test if the behavior changes if you use OpenCL backend.

**Comment by ebarjou** [comment]
I get the same behavior using OpenCL

**Comment by 9prady9** [comment]
thank you for checking that as well. I will look into it again tomorrow and get back to you.

**Comment by 9prady9** [comment]
@ebarjou Sorry about the delay.

I wasn't able to reproduce the level of 100% utilization on one core - I have set the affinity to one core, and there is momentary spike once, but that is it, it doesn't stay on 100% for that core and others cores doesn't even register this momentary spike as expected.
![Screenshot 2020-06-01 20 05 27](https://user-images.githubusercontent.com/3270458/83605755-14270380-a596-11ea-97a4-dd3742a89c47.png)


I have requested one of my colleague to try reproducing it on his dev setup. Will post again once he updates me.
